Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp

```diff
@@ -28,10 +28,103 @@
 #include "ArgumentCoders.h"
 #include "Decoder.h"
 #include "Encoder.h"
+#include "StreamConnectionEncoder.h"
 #include "Test.h"
 
 namespace TestWebKitAPI {
 
+// ArgumentCoderEncoderDecoderTest internally constructs the encoder object specified
+// as the template parameter, into which each test can encode the desired objects.
+// It then uses that Encoder's internal buffer to create a Decoder object, intended for
+// the test to decode those objects.
+
+struct EncoderDecoderTest {
+    static constexpr IPC::MessageName name()  { return static_cast<IPC::MessageName>(123); }
+};
+
+struct EncoderTypeNames {
+    template<typename T>
+    static std::string GetName(int)
+    {
+        if (std::is_same_v<T, IPC::Encoder>)
+            return "Encoder";
+        if (std::is_same_v<T, IPC::StreamConnectionEncoder>)
+            return "StreamConnectionEncoder";
+        return "<unknown>";
+    }
+};
+
+using EncoderTypes = ::testing::Types<IPC::Encoder, IPC::StreamConnectionEncoder>;
+
+template<typename T> class ArgumentCoderEncoderDecoderTest;
+
+template<>
+class ArgumentCoderEncoderDecoderTest<IPC::Encoder> : public ::testing::Test {
+public:
+    void SetUp() override
+    {
+        // IPC::Encoder's header data should fill out the first 16 bytes of the buffer.
+        m_encoder = makeUnique<IPC::Encoder>(EncoderDecoderTest::name(), 0);
+        ASSERT_EQ(m_encoder->bufferSize(), 16u);
+    }
+
+    IPC::Encoder& encoder() const { return *m_encoder; }
+    size_t headerSize() const { return 16; }
+    size_t encoderSize() const { return m_encoder->bufferSize(); }
+
+    std::unique_ptr<IPC::Decoder> createDecoder() const
+    {
+        return IPC::Decoder::create(m_encoder->buffer(), m_encoder->bufferSize(), { });
+    }
+
+private:
+    std::unique_ptr<IPC::Encoder> m_encoder;
+};
+
+template<>
+class ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder> : public ::testing::Test {
+public:
+    void SetUp() override
+    {
+        // IPC::StreamConnectionEncoder's header consists only of the 16-bit message name.
+        // To enable easier tests of proper alignment, a 64-bit value is enoded here to push
+        // the encoded size up to 16 bytes.
+        m_impl = makeUnique<Impl>();
+        ASSERT_EQ(m_impl->encoder.size(), 2u);
+        m_impl->encoder << uint64_t(0);
```

Comment:
stylistically this probably should still be static cast ..

Also, the intention is that everything works with the way the encoder is setup by default. E.g. the extra 64-bit value could be hiding some errors.

I think better strategy would be to not put the filler in. Rather write the tests and expect them to work.

If some test would need the encoder / size, you could adjust the size in the encoder size accessor. Like, `encodedDataSize()` or something. 

By quick finds, I couldn't determine if any tests in this patch would fail unintentionally due to the filler missing?

Of course, using the "encoded data size" is tricky, because the size of the alignment skip will depend on what is the first element being encoded / expected to be decoded. If you don't see a solution immediately, maybe we can ignore this suggestion and use the hardcoded filler.

Note: the uint64_t doesn't entirely circumvent the issue, as I belive max_align is larger than 16 bytes. E.g. there exist a type that induces a skip, so you'd encode the type and then assert a value larger than the type, which is surprising. However, the primary purpose of the filler, making StreamConnectionEncoder and Encoder invocations use the same value, would still be fulfilled.

Reply:
I can remove the filler. Subsequent tests can be rewritten to effectively manually round up and adjust the expected encoder size, which should be fine.

So the filler was used to accommodate the difference in header size between the two encoders, testing from a common encoder size. Funnily enough the `IPC::Encoder` header size was then increased a few days ago due to adoption of 128-bit identifiers.